### ISSUE TITLE
How to roll futures contracts in a long backtest?

### ISSUE THREAD

**Issue by xnox** [mementum/backtrader#132]
Hello,

I have a data feed of futures contracts, and i'm not sure how to roll them. Ideally something like close the contract on the close price of last day, and reopen with open price of the next or some such. Or simply adjust cash and the self.position.price "magically" as in practice it's not a trade that got initiated by the strategy. Any tips on how to best do that? Ideally without having to re-engineer the csv datafeed with extra bars inserted between the rollover dates.

Regards,

Dimitri

**Comment by mementum** [comment]
Hi,

This is a conceptual topic, because there is no definitive answer as to what's best.

Open literature from a commercial platform which also attempts to solve the problem: https://www.sierrachart.com/index.php?page=doc/ContinuousFuturesContractCharts.html

The best method, imho, would be one of the 2 *backadusted* methods. But this method is the best when the goal of the continuous future is to look at a chart. It is not well suited for a walk-forward backtesting analysis.

The best programming approach for a backtesting would be a data feed which accepts several inputs and has a *date* (or *volume*) rule to switch to the next feed at given moment in time. Defining *date rules* would need the addition of `dateutil` or a similar package as a dependency (at least for that data feed)

This may be an interesting addition to *backtrader*

**Comment by li-dp** [comment]
I've encountered the same probelms.

_Fake_ continuous future contract is not acceptable for me, because I was trying to backtest Calendar Spread Arbitrage.

So I modified backtrader brutely: 
1. load index market data as data0, because index data is continuous
2. load every future contract market data file, almost 100 files
3. when data0 has new tick, force cerebro call _next_, my stratergy's next will be called
4. in my stratergy,  I will determine which contract is tradeable according to data0's datatime.
5. send orders

By forcing cerebro call _next_, I can do my backtest, but I lost lot of backtrader's functionality: no  analyzers, no preload. Only the most import part, cash and position, works, that enough for me.
